components/bl-tour-guide-component/src/index.js

```diff
@@ -0,0 +1,111 @@
+import { useEffect, useState } from 'react';
+
+import Joyride, { ACTIONS, STATUS } from './lib/react-joyride.min';
+
+const { cn } = BackendlessUI.CSSUtils;
+
+export default function TourGuideComponent({ component, eventHandlers, elRef }) {
+  const { onChange } = eventHandlers;
+  const {
+    classList, display, style, controlVisibility, controlLabel, options, continuous, autoStart, closeOnEscape,
+    closeOnClickOverlay, scrolling, locale, scrollOffset, scrollDuration, scrollToFirstStep, spotlightClicks,
+  } = component;
+
+  const {
+    progressVisibility, skipButtonVisibility, overlayVisibility, backButtonVisibility, closeButtonVisibility,
+  } = component;
+
+  const {
+    spotlightPadding, beaconSize, arrowColor, backgroundColor, overlayColor, primaryColor, textColor, width,
+  } = component;
+
+  const tooltipStyles = { width, beaconSize, arrowColor, backgroundColor, overlayColor, primaryColor, textColor };
+
+  const [run, setRun] = useState(autoStart);
+  const [steps, setSteps] = useState(options);
+  const [helpers, setHelpers] = useState(null);
+  const [stepsKey, setStepsKey] = useState(0);
+
+  useEffect(() => {
+    setSteps(options);
+    setStepsKey(stepsKey + 1);
+  }, [options]);
+
+  const handleClickStart = event => {
+    event.preventDefault();
+
+    setRun(true);
+  };
+
+  const handleCallback = tourState => {
+    const { status, action } = tourState;
+
+    const isNextStepAction = action === ACTIONS.NEXT && status !== STATUS.FINISHED;
+    const isChangeStepAction = isNextStepAction || action === ACTIONS.PREV || action === ACTIONS.GO;
+
+    if (status === STATUS.FINISHED || status === STATUS.SKIPPED || action === ACTIONS.RESET) {
+      setRun(false);
+    }
+
+    if (continuous && isChangeStepAction) {
+      setTimeout(() => {
+        setRun(false);
+        setRun(true);
+      });
+    }
```

Comment:
I'm not clearly understand why we use this approach, but didn't controll `stepIndex`

Especially looks strange 51-54 lines where we use a timeout

Reply:
This approach is not associated with stepIndex. Setting a stepIndex will turn Joyride into controlled mode, which will require additional logic that is currently being executed by the library.

According to the docs, need to set our running state to false, so we can restart if we click start again.:
```
    if (status === STATUS.FINISHED || status === STATUS.SKIPPED || action === ACTIONS.RESET) {
      setRun(false);
    }
```


As for this logic:
```
    const isNextStepAction = action === ACTIONS.NEXT && status !== STATUS.FINISHED;
    const isChangeStepAction = isNextStepAction || action === ACTIONS.PREV || action === ACTIONS.GO;

    if (continuous && isChangeStepAction) {
      setTimeout(() => {
        setRun(false);
        setRun(true);
      });
    }
```
The library has a bug with the tooltip appearance in some cases of using next, prev and goTo helpers. To fix this issue, we need to update the tour state.
We use timeout because we need to update the state after the current callback is completed.

Reply:
Is controlled mode requires a lot of logic to implement? If not - it will be good for the future.

If we leave this approach - please add comments on why we do it(also if the library has a known issue - add this link too)

That is good practice for all unobvious solutions.